[dynamo] add torch.compiler.set_stance

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137504

Attempt # 2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_stance` function that can force `torch.compile` regions to run eagerly.

See added tests for usage examples.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

williamwen42 committed:
[dynamo] add torch.compiler.set_phase

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -114,6 +114,54 @@ def _maybe_set_eval_frame(callback: DynamoCallback):
+_set_phase._dynamo_forbidden = True  # type: ignore[attr-defined]
+
+
+class SetPhaseCtxMgrDecorator:
```
Using manual context manager/decorator class instead of `contextlib.contextmanager` in order to mark a decorated function as forbidden in graph.

williamwen42 committed:
Update on "[dynamo] add torch.compiler.set_phase"


Attempt #2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_phase` function that can force `torch.compile` regions to run eagerly. Other options (such as "fail_on_recompile" and "eager_on_recompile") to be added in a followup PR.

See added tests for usage examples.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch.compiler.set_phase"


Attempt #2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_phase` function that can force `torch.compile` regions to run eagerly. Other options (such as "fail_on_recompile" and "eager_on_recompile") to be added in a followup PR.

See added tests for usage examples.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -114,6 +114,54 @@ def _maybe_set_eval_frame(callback: DynamoCallback):
+
+
+# forbid in graph
+_set_phase._dynamo_forbidden = True  # type: ignore[attr-defined]
```
Flag is set, e.g. here https://github.com/pytorch/pytorch/blob/38027b9b4776f3fc3034193b28b7f80b58afb5de/torch/_dynamo/decorators.py#L168 to implement `forbid_in_graph`.

It is checked here https://github.com/pytorch/pytorch/blob/38027b9b4776f3fc3034193b28b7f80b58afb5de/torch/_dynamo/trace_rules.py#L3085

yf225 replied:
ah I see, thanks!

jansel commented on torch/compiler/__init__.py:
```diff
@@ -228,6 +229,25 @@ def disable(fn=None, recursive=True):
     return torch._dynamo.disable(fn, recursive)
 
 
+def set_phase(phase: str):
```
`phase` seems like the wrong name here.  Forcing eager does not seem like a compilation phase.

There also uses of this that don't have anything to do with phases.

jansel commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -819,6 +819,24 @@ static PyObject* set_eval_frame_py(PyObject* dummy, PyObject* callback) {
   return set_eval_frame(callback, PyThreadState_GET());
 }
 
+static PyObject* set_eval_frame_callback_enabled(PyObject* dummy, PyObject* enabled) {
+  bool prior = eval_frame_callback_enabled_get();
+  if (enabled == Py_True) {
+    eval_frame_callback_enabled_set(true);
+  } else if (enabled == Py_False) {
+    eval_frame_callback_enabled_set(false);
+  } else {
+    DEBUG_TRACE0("arg error");
+    PyErr_SetString(PyExc_TypeError, "expected bool");
+    return NULL;
+  }
+  return PyBool_FromLong(prior);
+}
```
Doesn't this do the same thing as `set_eval_frame(None)` (which disables dynamo and makes everything run eagerly).

We also have `TORCHDYNAMO_DISABLE=1` which does something similar.

The one advantage this has is it is thread local, but IMO we should just fix the thread safety of the existing stuff rather than adding a duplicate API.

williamwen42 replied:
`eval_frame_callback_enabled_get` is higher priority than `set_eval_frame`. The new API is useful for cases where we want to ignore future `torch.compile` invocations, which would overwrite the eval frame handler.

We have a lot of discussion over at https://github.com/pytorch/pytorch/pull/132926.

jansel replied:
It would be much faster to disable frame evaluation entirely (do the check once when we hit torch.compiled code), since then the CPython JIT can kick in.  We also won't need to do an extra check on every function call.

jansel commented:
Also note that a version of `eager_on_recompile` already exists with `torch._dynamo.run` (maps to `set_eval_frame(True)`), so no need to recreate that.

Similarly, `fail_on_recompile` can be done by setting the backend to one that raises an error (without any changes to eval_frame.c).

williamwen42 committed:
Update on "[dynamo] add torch.compiler.set_phase"


Attempt #2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_phase` function that can force `torch.compile` regions to run eagerly. Other options (such as "fail_on_recompile" and "eager_on_recompile") to be added in a followup PR.

See added tests for usage examples.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch.compiler.set_stance"


Attempt #2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_stance` function that can force `torch.compile` regions to run eagerly.

See added tests for usage examples.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ezyang commented:
I'm going to defer to @jansel on this since he has more opinions on the internal implementation details, but holler if you need me to unblock you

jansel commented on torch/_dynamo/decorators.py:
```diff
@@ -81,6 +82,36 @@ def skip(fn=None):
+
+    def __init__(self, stance: str) -> None:
+        self.stance = stance
+        self.prev = _set_stance(stance)
```
Won't calling this in the constructor mean that, defining a function:
```
@set_stance("foo")
def bar():
   pass
```
Will change the global stance even if `bar()` is unused?

williamwen42 replied:
`__call__` will run in decorator mode, which resets the stance state to the previous state (this logic was taken from the `torch.set_grad_enabled` implementation).

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -114,6 +114,43 @@ def _maybe_set_eval_frame(callback: DynamoCallback):
+                "Detected recompile when torch.compile stance is 'fail_on_recompile'"
+            )
+
+        fail_callback._torchdynamo_orig_callable = callback  # type: ignore[attr-defined]
```
Add a comment about why this is needed.

jansel commented on torch/compiler/__init__.py:
```diff
@@ -228,6 +229,26 @@ def disable(fn=None, recursive=True):
 
 
+def set_stance(stance: str):
+    """
+    Set the current stance of the compiler.
+    Can be used as a function, context manager, or decorator.
+    Do not use this function inside a `torch.compile` region - an error will be raised otherwise.
+
+    Args:
+        stance: The stance to set the compiler to. Valid values are:
+            - "default": The default stance, used for normal compilation.
+            - "force_eager": Ignore all `torch.compile` directives.
+    """
```
This needs to be updated.  Let's also improve it with some example usage.

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -114,6 +114,43 @@ def _maybe_set_eval_frame(callback: DynamoCallback):
+def _set_stance(stance: str):
+    global _stance
+    prior = _stance
+    _stance = stance
```
Should we assert here that there isn't currently an eval frame hook set?

If there is, then the stance won't take effect until the next dynamo region is entered.

williamwen42 replied:
`_dynamo_forbidden` prevents `set_stance` from being used in a `torch.compile` region. Was there another case you were thinking about?

jansel replied:
Calling this function after a graph break that forces a fallback to eager.

williamwen42 committed:
Update on "[dynamo] add torch.compiler.set_stance"


Attempt # 2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_stance` function that can force `torch.compile` regions to run eagerly.

See added tests for usage examples.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch.compiler.set_stance"


Attempt # 2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_stance` function that can force `torch.compile` regions to run eagerly.

See added tests for usage examples.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -114,6 +115,64 @@ def _maybe_set_eval_frame(callback: DynamoCallback):
+
+        return callback
+
+    if _stance.backend is not None:
+        raise RuntimeError("non-default stance cannot have force_backend set")
```
We could do this check inside `set_stance` (since that is run less often than this one)

williamwen42 committed:
Update on "[dynamo] add torch.compiler.set_stance"


Attempt # 2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_stance` function that can force `torch.compile` regions to run eagerly.

See added tests for usage examples.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add torch.compiler.set_stance"


Attempt # 2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_stance` function that can force `torch.compile` regions to run eagerly.

See added tests for usage examples.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

williamwen42 commented:
@pytorchbot merge

williamwen42 commented:
@pytorchbot merge

williamwen42 commented:
@ezyang it's already implemented in this PR, although I overlooked adding a test and including it as a usage example. I can add this in a followup.

yf225 commented:
@williamwen42 also I've started using the direct-mutating API: https://github.com/pytorch/pytorch/pull/138113/files and it's awesome :)
